src/pages/storage/MigrateVolumeBtn.tsx

```diff
@@ -0,0 +1,137 @@
+import { LxdStorageVolume } from "types/storage";
+import MigrateVolumeModal from "./MigrateVolumeModal";
+import { migrateStorageVolume } from "api/storage-pools";
+import { useLocation, useNavigate } from "react-router-dom";
+
+interface Props {
+  storageVolume: LxdStorageVolume;
+  project: string;
+  classname?: string;
+  onClose?: () => void;
+}
+
+const MigrateVolumeBtn: FC<Props> = ({
+  storageVolume,
+  project,
+  classname,
+  onClose,
+}) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const { openPortal, closePortal, isOpen, Portal } = usePortal();
+  const queryClient = useQueryClient();
+  const navigate = useNavigate();
+  const [isVolumeLoading, setVolumeLoading] = useState(false);
+  const location = useLocation();
+
+  const handleSuccess = (
+    newTarget: string,
+    storageVolume: LxdStorageVolume,
+  ) => {
+    toastNotify.success(
+      <>
+        Volume <ItemName item={{ name: storageVolume.name }} bold />{" "}
+        successfully migrated to pool{" "}
+        <ItemName item={{ name: newTarget }} bold />
+      </>,
+    );
+
+    const oldVolumeUrl = `/ui/project/${storageVolume.project}/storage/pool/${storageVolume.pool}/volumes/${storageVolume.type}/${storageVolume.name}`;
+    const newVolumeUrl = `/ui/project/${storageVolume.project}/storage/pool/${newTarget}/volumes/${storageVolume.type}/${storageVolume.name}`;
+    if (location.pathname.startsWith(oldVolumeUrl)) {
+      navigate(newVolumeUrl);
+    }
+  };
+
+  const notifyFailure = (
+    e: unknown,
+    storageVolume: string,
+    targetPool: string,
+  ) => {
+    setVolumeLoading(false);
+    toastNotify.failure(
+      `Migration failed for volume ${storageVolume} to pool ${targetPool}`,
+      e,
+    );
+  };
+
+  const handleFailure = (
+    msg: string,
+    storageVolume: string,
+    targetPool: string,
+  ) => {
+    notifyFailure(new Error(msg), storageVolume, targetPool);
+  };
+
+  const handleFinish = () => {
+    void queryClient.invalidateQueries({
+      queryKey: [queryKeys.storage, storageVolume.name],
+    });
+    setVolumeLoading(false);
+  };
+
+  const handleMigrate = (targetPool: string) => {
+    setVolumeLoading(true);
+    migrateStorageVolume(storageVolume, targetPool)
+      .then((operation) => {
+        eventQueue.set(
+          operation.metadata.id,
+          () => handleSuccess(targetPool, storageVolume),
+          (err) => handleFailure(err, storageVolume.name, targetPool),
+          handleFinish,
+        );
+        toastNotify.info(
+          `Migration started for volume ${storageVolume.name} to pool ${targetPool}`,
+        );
+        void queryClient.invalidateQueries({
+          queryKey: [queryKeys.storage, storageVolume.name, project],
+        });
+      })
+      .catch((e) => {
+        notifyFailure(e, storageVolume.name, targetPool);
+      })
+      .finally(() => {
+        handleClose();
```

Comment:
I think we should not close on failure, the user might want to retry and we should keep the modal open. Only in the success case we should call handleClose.

Reply:
The success case, handleSuccess, doesn't need to handleClose in this case, right?

Reply:
Yeah, we'd close already on submitting the operation, no need to close again once the operation succeeds :)

Reply:
I might be overseeing, but don't we miss to close now in the success case? Like when the info notification gets submitted.

Reply:
This is addressed be a new comment below.

Reply:
i'm not sure I understand, but yes, there is no reference to handleClose or Close when the success criteria is met.